src/subword.rs

```diff
@@ -1,14 +1,99 @@
 use std::cmp;
 use std::collections::VecDeque;
+use std::fmt;
 use std::hash::{Hash, Hasher};
+use std::marker::PhantomData;
 use std::ops::Deref;
 
 use fnv::FnvHasher;
 
 use crate::util::CollectWithCapacity;
 
+/// N-Gram indexer
+///
+/// An indexer maps an n-gram to an index in the subword embedding
+/// matrix.
+pub trait Indexer {
+    /// Map an n-gram to an index in the subword embedding matrix.
+    fn index(&self, ngram: &StrWithCharLen) -> u64;
+}
+
+/// Indexer using a hash function.
+///
+/// This indexer first hashes a given n-gram and then maps the
+/// resulting hash into *2^buckets_exp* buckets.
+///
+/// The largest possible bucket exponent is 64.
+pub struct HashIndexer<H> {
+    mask: u64,
+    _phantom: PhantomData<H>,
+}
+
+impl<H> HashIndexer<H> {
+    /// Construct an `FNV1Indexer`.
+    ///
+    /// The largest possible bucket exponent is 64.
+    pub fn new(buckets_exp: usize) -> Self {
+        assert!(
+            buckets_exp <= 64,
+            "The largest possible buckets exponent is 64."
+        );
+
+        let mask = if buckets_exp == 64 {
+            !0
+        } else {
+            (1 << buckets_exp) - 1
+        };
+
+        HashIndexer {
+            mask,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<H> Clone for HashIndexer<H> {
+    fn clone(&self) -> Self {
+        HashIndexer {
+            mask: self.mask,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<H> Copy for HashIndexer<H> {}
+
+impl<H> fmt::Debug for HashIndexer<H> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        // std::intrinsics::type_name requires nightly
+        write!(f, "HashIndexer<impl Hasher> {{ mask: {} }}", self.mask)
+    }
+}
+
+impl<H> Eq for HashIndexer<H> {}
+
+impl<H> Indexer for HashIndexer<H>
+where
+    H: Default + Hasher,
+{
+    fn index(&self, ngram: &StrWithCharLen) -> u64 {
```

Comment:
There's a collision with `Index::index`, I'm not sure whether this could lead to any confusion

Reply:
Or this could be implemented through `Index::index`. In case of an actual ngram lookup, that would even be more fitting. In case of only using the bucket it's also not too far fetched.

Reply:
> There's a collision with `Index::index`, I'm not sure whether this could lead to any confusion

I think normally it is not a problem, but since is so prominent, I renamed the method to `index_ngram` just to be certain.

Reply:
> Or this could be implemented through `Index::index`. In case of an actual ngram lookup, that would even be more fitting. In case of only using the bucket it's also not too far fetched.

That would be very elegant, but the problem is that `Index::index` is required to return a reference:

https://doc.rust-lang.org/std/ops/trait.Index.html#tymethod.index

which we can't :(.

Reply:
> That would be very elegant, but the problem is that `Index::index` is required to return a reference:

Ah, I forgot about that. Unfortunate!